File: src/delta_review/gitlab/positions.py
from __future__ import annotations

from copy import deepcopy
from dataclasses import dataclass
from hashlib import sha1
from typing import Any


@dataclass(frozen=True)
class Version:
    base_sha: str
    start_sha: str
    head_sha: str


@dataclass(frozen=True)
class DiffSelection:
    old_path: str
    new_path: str
    start_old: int | None
    start_new: int | None
    end_old: int | None
    end_new: int | None


def _line_code(path: str, old: int | None, new: int | None) -> str:
    digest = sha1(path.encode()).hexdigest()
    return f"{digest}_{old or 0}_{new or 0}"


def _line_type(
    old: int | None, new: int | None
) -> str | None:
    if old is not None and new is not None:
        return None
    return "new" if new is not None else "old"


def _range_endpoint(
    path: str, old: int | None, new: int | None
) -> dict[str, str | int | None]:
    endpoint: dict[str, str | int | None] = {
        "line_code": _line_code(path, old, new),
        "type": _line_type(old, new),
    }
    if old is not None:
        endpoint["old_line"] = old
    if new is not None:
        endpoint["new_line"] = new
    return endpoint


def build_position(
    selection: DiffSelection, version: Version
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "position_type": "text",
        "base_sha": version.base_sha,
        "start_sha": version.start_sha,
        "head_sha": version.head_sha,
        "old_path": selection.old_path,
        "new_path": selection.new_path,
    }
    if selection.end_old is not None:
        payload["old_line"] = selection.end_old
    if selection.end_new is not None:
        payload["new_line"] = selection.end_new

    if (selection.start_old, selection.start_new) != (
        selection.end_old,
        selection.end_new,
    ):
        path = selection.new_path or selection.old_path
        payload["line_range"] = {
            "start": _range_endpoint(
                path, selection.start_old, selection.start_new
            ),
            "end": _range_endpoint(
                path, selection.end_old, selection.end_new
            ),
        }
    return payload


def build_legacy_position(
    selection: DiffSelection, version: Version
) -> dict[str, Any]:
    payload = deepcopy(build_position(selection, version))
    for key in ("old_line", "new_line"):
        if key in payload and payload[key] is not None:
            payload[key] = str(payload[key])

    line_range = payload.get("line_range")
    if not isinstance(line_range, dict):
        return payload
    for name in ("start", "end"):
        endpoint = line_range[name]
        for key in ("old_line", "new_line"):
            value = endpoint.get(key)
            endpoint[key] = None if value is None else str(value)
    return payload

File: web/src/gitlab/positions.ts
import type { BackendSelection } from '../review/selection'

export type DiffSelection = BackendSelection

export interface Version {
  base_sha: string
  start_sha: string
  head_sha: string
}

async function sha1Hex(value: string): Promise<string> {
  const digest = await crypto.subtle.digest(
    'SHA-1',
    new TextEncoder().encode(value),
  )
  return Array.from(new Uint8Array(digest))
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('')
}

async function lineCode(
  path: string,
  oldLine: number | null,
  newLine: number | null,
): Promise<string> {
  return `${await sha1Hex(path)}_${oldLine ?? 0}_${newLine ?? 0}`
}

function lineType(
  oldLine: number | null,
  newLine: number | null,
): 'old' | 'new' | null {
  if (oldLine !== null && newLine !== null) return null
  return newLine !== null ? 'new' : 'old'
}

async function rangeEndpoint(
  path: string,
  oldLine: number | null,
  newLine: number | null,
): Promise<Record<string, unknown>> {
  const endpoint: Record<string, unknown> = {
    line_code: await lineCode(path, oldLine, newLine),
    type: lineType(oldLine, newLine),
  }
  if (oldLine !== null) endpoint.old_line = oldLine
  if (newLine !== null) endpoint.new_line = newLine
  return endpoint
}

export async function buildPosition(
  selection: DiffSelection,
  version: Version,
): Promise<Record<string, unknown>> {
  const payload: Record<string, unknown> = {
    position_type: 'text',
    base_sha: version.base_sha,
    start_sha: version.start_sha,
    head_sha: version.head_sha,
    old_path: selection.old_path,
    new_path: selection.new_path,
  }
  if (selection.end_old !== null) payload.old_line = selection.end_old
  if (selection.end_new !== null) payload.new_line = selection.end_new

  const isMultiline =
    selection.start_old !== selection.end_old ||
    selection.start_new !== selection.end_new
  if (isMultiline) {
    const path = selection.new_path || selection.old_path
    payload.line_range = {
      start: await rangeEndpoint(
        path,
        selection.start_old,
        selection.start_new,
      ),
      end: await rangeEndpoint(path, selection.end_old, selection.end_new),
    }
  }
  return payload
}

export async function buildLegacyPosition(
  selection: DiffSelection,
  version: Version,
): Promise<Record<string, unknown>> {
  const payload = await buildPosition(selection, version)
  for (const key of ['old_line', 'new_line'] as const) {
    const value = payload[key]
    if (value !== undefined && value !== null) payload[key] = String(value)
  }

  const lineRange = payload.line_range as
    | Record<'start' | 'end', Record<string, unknown>>
    | undefined
  if (!lineRange) return payload
  for (const name of ['start', 'end'] as const) {
    const endpoint = lineRange[name]
    for (const key of ['old_line', 'new_line'] as const) {
      const value = endpoint[key]
      endpoint[key] = value === undefined || value === null ? null : String(value)
    }
  }
  return payload
}
